v2/internal/controllers/crd_machinelearningservices_test.go

```diff
@@ -0,0 +1,288 @@
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func Test_MachineLearning_Workspaces_CRUD(t *testing.T) {
+	t.Parallel()
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+
+	sa := newStorageAccount(tc, rg)
+
+	tc.CreateResourceAndWait(sa)
+
+	kv := newVault(tc, rg)
+	tc.CreateResourceAndWait(kv)
+
+	workspace := newWorkspace(tc, testcommon.AsOwner(rg), sa, kv, tc.AzureRegion)
+
+	tc.CreateResourcesAndWait(workspace)
+
+	// There should be no secrets at this point
+	list := &v1.SecretList{}
+	tc.ListResources(list, client.InNamespace(tc.Namespace))
+	tc.Expect(list.Items).To(gomega.HaveLen(0))
+
+	tc.RunParallelSubtests(
+		testcommon.Subtest{
+			Name: "Test_WorkspacesCompute_CRUD",
+			Test: func(testContext *testcommon.KubePerTestContext) {
+				WorkspaceCompute_CRUD(tc, testcommon.AsOwner(workspace), rg)
+			},
+		},
+	)
+
+	tc.DeleteResourcesAndWait(workspace, kv, sa, rg)
+
+}
+
+func Workspaces_WriteSecrets(tc *testcommon.KubePerTestContext, workspace *machinelearningservices.Workspace) {
+	old := workspace.DeepCopy()
+	workspaceKeysSecret := "workspacekeyssecret"
+	workspace.Spec.OperatorSpec = &machinelearningservices.WorkspaceOperatorSpec{
+		Secrets: &machinelearningservices.WorkspaceOperatorSecrets{
+			PrimaryNotebookAccessKey:   &genruntime.SecretDestination{Name: workspaceKeysSecret, Key: "primaryNotebookAccessKey"},
+			SecondaryNotebookAccessKey: &genruntime.SecretDestination{Name: workspaceKeysSecret, Key: "secondaryNotebookAccessKey"},
+			UserStorageKey:             &genruntime.SecretDestination{Name: workspaceKeysSecret, Key: "userStorageKey"},
+		},
+	}
+	tc.PatchResourceAndWait(old, workspace)
+
+	tc.ExpectSecretHasKeys(
+		workspaceKeysSecret,
+		"primaryNotebookAccessKey",
+		"secondaryNotebookAccessKey",
+		"userStorageKey")
+}
+
+func newWorkspace(tc *testcommon.KubePerTestContext, owner *genruntime.KnownResourceReference, sa *storage.StorageAccount, kv *v1beta20210401preview.Vault, location *string) *machinelearningservices.Workspace {
+	identityType := machinelearningservices.IdentityTypeSystemAssigned
+
+	workspaces := &machinelearningservices.Workspace{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.NoSpaceNamer.GenerateName("work")),
+		Spec: machinelearningservices.Workspaces_Spec{
+			Location: location,
+			Owner:    owner,
+			Sku: &machinelearningservices.Sku{
+				Name: to.StringPtr("Standard_S1"),
+				Tier: to.StringPtr("Basic"),
+			},
+			AllowPublicAccessWhenBehindVnet: to.BoolPtr(false),
+			Identity: &machinelearningservices.Identity{
+				Type: &identityType,
+			},
+			StorageAccountReference: tc.MakeReferenceFromResource(sa),
+			KeyVaultReference:       tc.MakeReferenceFromResource(kv),
+		},
+	}
+	return workspaces
+}
+
+func Test_MachineLearning_WorkspaceConnection_CRUD(t *testing.T) {
```

Comment:
Do you actually need this test? It seems like it's just creating another workspace and then testing connection in that.

Could you instead have a structure like this?

- Test_MachineLearning_Workspaces_CRUD
  - Parallel:
    - WorkspaceCompute_CRUD
    - WorkspaceConnection_CRUD
  - Sequential:
    - Workspaces_WriteSecrets

Since you already have the workspace for the first test just wondering if it can be reused?

Reply:
Ah true, Had to do this way earlier when I mentioned about getting the race issues while recording the tests.

Reply:
Done!